src/analyses/ai.h

```diff
@@ -406,70 +416,36 @@ class ait:public ai_baset
 public:
   // constructor
   ait()
-    : ai_baset(),
-      domain_factory(
-        util_make_unique<ai_domain_factory_default_constructort<domainT>>())
+    : ai_baset(
+        util_make_unique<ai_domain_factory_default_constructort<domainT>>(),
+        util_make_unique<location_sensitive_storaget>())
   {
   }
 
   explicit ait(std::unique_ptr<ai_domain_factory_baset> &&df)
-    : ai_baset(std::move(df))
+    : ai_baset(std::move(df), util_make_unique<location_sensitive_storaget>())
   {
   }
 
   typedef goto_programt::const_targett locationt;
 
   /// Find the analysis result for a given location.
+  // The older interface for non-modifying access
+  // Not recommended as it will throw an exception if a location has not
+  // been reached in an analysis and there is no (other) way of telling
+  // if a location has been reached.
+  DEPRECATED(SINCE(2019, 08, 01, "use abstract_state_{before,after} instead"))
   const domainT &operator[](locationt l) const
   {
-    typename state_mapt::const_iterator it=state_map.find(l);
-    if(it==state_map.end())
-      throw std::out_of_range("failed to find state");
-
-    return static_cast<const domainT &>(*(it->second));
-  }
-
-  /// Used internally by the analysis.
-  std::shared_ptr<const statet>
-  abstract_state_before(locationt t) const override
-  {
-    typename state_mapt::const_iterator it = state_map.find(t);
-    if(it == state_map.end())
-      return domain_factory->make_domain(t);
+    auto p(storage->abstract_state_before(l, *domain_factory));
```

Comment:
Could you change this to `auto p = storage->....`? or `auto p{storage->....}`? With the combination of `auto` and the parentheses I had to do a double take here.

Also what's `p`?

Reply:
Now `auto p = storage->abstract_state_before(l, *domain_factory);`.  Better?